### DISCUSSION TITLE
New search is less helpful than old search for complete matches of substrings

### DISCUSSION THREAD

**Discussion by jsoref** [community/community#41709] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Product Feedback

### Body

This search finds precisely what I want:
https://github.com/search?type=codelegacy&q=repo%3Adart-lang%2Fsdk+roduct

<img width="661" alt="image" src="https://user-images.githubusercontent.com/2119212/207558510-09f435c9-a473-4f13-b9c0-e1d82193a25c.png">

This search technically does but practically doesn't:
https://github.com/search?type=code&q=repo%3Adart-lang%2Fsdk+roduct

specifically the results are so cluttered that my eyes miss that the first hit *actually* has the hit I want:
https://github.com/dart-lang/sdk/blob/434b65470b0c52a9d4015a9d8810e2f74237604e/runtime/vm/flag_list.h#L75

because my eyes scan the first block, see:
https://github.com/dart-lang/sdk/blob/434b65470b0c52a9d4015a9d8810e2f74237604e/runtime/vm/flag_list.h#L33

and conclude "oh, gosh, it's a substring match, this block is useless, let's check the next block".

<img width="697" alt="image" src="https://user-images.githubusercontent.com/2119212/207558461-318d9c1e-90ce-4c7a-b755-eb9a147ad13e.png">

One way to save me from this would be to use a different color for perfect matches vs substring matches (probably making substring matches a paler color).

**Comment by look** [comment, marked as answer]
Thanks for the feedback @jsoref. I like the idea of indicating exact matches better. We could also use it in our ranking.

For a technical explanation, the reason why the old search finds "roduct" so well is that it tokenizes whole words and some programming-related separators (like `-`, `_`, and case changes such as `productManager` turning into `product`, `manager`). The new search supports substring matches, which is why you're getting so many additional results. One way to help is to be more precise in what you're searching for. For example, [this regex search for `/\s+roduct\s+/`](https://github.com/search?type=code&q=repo%3Adart-lang%2Fsdk+%2F%5Cs%2Broduct%5Cs%2B%2F) finds only the result you were looking for.

**Reply by jsoref** [reply]
Thanks 😄 .

I've worked on a cross reference (MXR) and am currently working on a [spell-checker](https://github.com/check-spelling/check-spelling) which is aware of programmer conventions (ala the old engine). It certainly felt like your engine was effectively moving from the MXR model to the OpenGrok model which was pure substring. I can understand why that would be appealing to end users, but, I certain would encourage improving the scoring as described here.

From memory the new engine has tended to be fairly slow/annoyed when I only give it a regex. (I can't tell if that's happening here, as I'm uncertain if it's cheating since you recently searched for it..., and I don't have a particularly good sense of the size of the dart repository, from my perspective, it's a medium size repository -- even though I'm contributing in bite-sized chunks.)

**Reply by look** [reply]
Thanks for sharing what you're working on, the spell checker sounds interesting!

Our main focus right now is performance and scaling (so we can release the the search to everyone!) so if there are regexes you run into that are slow, definitely let us know. They are slower to evaluate than exact match searches, but a fairly specific regex within a single repo should be pretty quick.

**Reply by jsoref** [reply]
Ah, that's it, I'm often not constraining the searches to a single repository. I think I'm generally collecting/vetting patterns:

https://github.com/check-spelling/spell-check-this/blob/prerelease/.github/actions/spelling/candidate.patterns
https://github.com/check-spelling/spell-check-this/blob/prerelease/.github/actions/spelling/line_forbidden.patterns

---

Right now I'm trying to tune file extensions:
https://github.com/check-spelling/spell-check-this/blob/prerelease/.github/actions/spelling/excludes.txt

initially I ran:
https://github.com/search?q=path%3A.lock&type=code

... which seemed like it was doing the right thing, until I realized that the `.` was being ignored, which wasn't very nice of it...

... my final url before I stopped was https://github.com/search?q=path%3A%2F%5C.lock%24%2F+NOT+language%3AGemfile.lock+NOT+language%3AElixir+NOT+language%3AErlang+NOT+language%3ARuby+NOT+language%3AOrg&type=code&l=Org -- shortly after I realized that I'd already crossed that bridge (oh well).

**Reply by look** [reply]
> ... which seemed like it was doing the right thing, until I realized that the `.` was being ignored, which wasn't very nice of it...

For this one, we have a special glob syntax you can use: [`path:*.lock`](https://github.com/search?q=path%3A*.lock&type=code)

**Reply by jsoref** [reply]
If it looks like i'm trying to search for that, it'd be nice if GitHub suggested it (it does occasionally make suggestions).

**Comment by Houston256** [comment]
Also a problem when searching issues, whole-word matches work perfectly, substring matches basically never work.
